core/bn.js

```diff
@@ -290,14 +286,177 @@ sjcl.bn.prototype = {
 
   /** this ^ x mod N */
   powermod: function(x, N) {
-    var result = new sjcl.bn(1), a = new sjcl.bn(this), k = new sjcl.bn(x);
-    while (true) {
-      if (k.limbs[0] & 1) { result = result.mulmod(a, N); }
-      k.halveM();
-      if (k.equals(0)) { break; }
-      a = a.mulmod(a, N);
-    }
-    return result.normalize().reduce();
+    x = new sjcl.bn(x);
+    N = new sjcl.bn(N);
+
+    // Jump to montpowermod if possible.
+    if ((N.limbs[0] & 1) == 1) {
+      var montOut = this.montpowermod(x, N);
+
+      if (montOut != false) { return montOut; } // else go to slow powermod
+    }
+
+    var i, j, l = x.normalize().trim().limbs, out = new this._class(1), pow = this;
+
+    for (i=0; i<l.length; i++) {
+      for (j=0; j<this.radix; j++) {
+        if (l[i] & (1<<j)) { out = out.mulmod(pow, N); }
+        if (i == (l.length - 1) && l[i]>>(j + 1) == 0) { break; }
+
+        pow = pow.mulmod(pow, N);
+      }
+    }
+
+    return out;
+  },
+
+  /** this ^ x mod N with Montomery reduction */
+  montpowermod: function(x, N) {
+    x = new sjcl.bn(x).normalize().trim();
+    N = new sjcl.bn(N);
+
+    var i, j,
+      radix = this.radix,
+      out = new this._class(1),
+      pow = this.copy();
+
+    // Generate R as a cap of N.
+    var R, s, wind, bitsize = x.bitLength();
+
+    R = new sjcl.bn({
+      limbs: N.copy().normalize().trim().limbs.map(function() { return 0; })
+    });
+
+    for (s = this.radix; s > 0; s--) {
+      if (((N.limbs[N.limbs.length - 1] >> s) & 1) == 1) {
+        R.limbs[R.limbs.length - 1] = 1 << s;
+        break;
+      }
+    }
+
+    // Calculate window size as a function of the exponent's size.
+    if (bitsize == 0) { return this;
+    } else if (bitsize < 18)  { wind = 1;
+    } else if (bitsize < 48)  { wind = 3;
+    } else if (bitsize < 144) { wind = 4;
+    } else if (bitsize < 768) { wind = 5;
+    } else { wind = 6; }
+
+    // Find R' and N' such that R * R' - N * N' = 1.
+    var RR = R.copy(), NN = N.copy(), RP = new sjcl.bn(1), NP = new sjcl.bn(0), RT = R.copy();
+
+    while (RT.greaterEquals(1)) {
+      RT.halveM();
+
+      if ((RP.limbs[0] & 1) == 0) {
+        RP.halveM();
+        NP.halveM();
+      } else {
+        RP.addM(NN);
+        RP.halveM();
+
+        NP.halveM();
+        NP.addM(RR);
+      }
+    }
+
+    RP = RP.normalize();
+    NP = NP.normalize();
+
+    RR.doubleM()
+    var R2 = RR.mulmod(RR, N);
+
+    // Check whether the invariant holds.
+    // If it doesn't, we can't use Montgomery reduction on this modulus.
+    if (!RR.mul(RP).sub(N.mul(NP)).equals(1)) {
+      return false;
+    }
+
+    var montIn = function(c) { return montMul(c, R2) },
```

Comment:
@Nilos There's not really a clean way to do that... Those functions use the calculation of `R` and `N` from above.  If they were pulled out, you'd have to either pass in both which is weird because `R` isn't an independent parameter, it's directly dependent on `N`.  Just passing in `N` would mean wasting time re-calculating `R`.

Should I just go with the creating-a-new-class idea?